packages/component-base/src/data-provider-controller/data-provider-controller.d.ts

```diff
@@ -0,0 +1,154 @@
+/**
+ * @license
+ * Copyright (c) 2021 - 2023 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+import type { ReactiveController } from 'lit';
+import type { Cache } from './cache.js';
+
+type DataProviderDefaultParams = {
+  page: number;
+  pageSize: number;
+  parentItem?: unknown;
+};
+
+export type DataProviderCallback = (items: unknown[], size?: number) => void;
+
+export type DataProvider<DataProviderParams extends Record<string, unknown>> = (
+  params: DataProviderDefaultParams & DataProviderParams,
+  callback: DataProviderCallback,
+) => void;
+
+/**
+ * A controller that stores and manages items loaded with a data provider.
+ */
+export class DataProviderController<DataProviderParams extends Record<string, unknown>> implements ReactiveController {
+  /**
+   * The controller host element.
+   */
+  host: HTMLElement;
+
+  /**
+   * A callback that returns data based on the passed params such as
+   * `page`, `pageSize`, `parentItem`, etc.
+   */
+  dataProvider: DataProvider<DataProviderParams>;
+
+  /**
+   * A callback that returns additional params that need to be passed
+   * to the data provider callback with every request.
+   */
+  dataProviderParams: () => DataProviderParams;
+
+  /**
+   * A number of items in the root cache.
+   */
+  size?: number;
+
+  /**
+   * A number of items to display per page.
+   */
+  pageSize: number;
+
+  /**
+   * A callback that returns whether the given item is expanded.
+   */
+  isExpanded: (item: unknown) => boolean;
+
+  /**
+   * A reference to the root cache instance.
+   */
+  rootCache: Cache;
+
+  constructor(
+    host: HTMLElement,
+    config: {
+      size?: number;
+      pageSize: number;
+      isExpanded(item: unknown): boolean;
+      dataProvider: DataProvider<DataProviderParams>;
```

Comment:
The ordering here (and in the test file) is off:

<img width="553" alt="Screenshot 2023-09-25 at 16 44 45" src="https://github.com/vaadin/web-components/assets/1222264/0e9e831d-5cc6-4234-8ddb-22a92d5dc393">